dask/dataframe/core.py

```diff
@@ -5727,62 +5728,68 @@ def apply_concat_apply(
         split_out_setup_kwargs,
     )
 
-    # Chunk
-    a = f"{token or funcname(chunk)}-chunk-{token_key}"
-    if len(args) == 1 and isinstance(args[0], _Frame) and not chunk_kwargs:
-        dsk = {
-            (a, 0, i, 0): (chunk, key) for i, key in enumerate(args[0].__dask_keys__())
-        }
-    else:
-        dsk = {
-            (a, 0, i, 0): (
-                apply,
-                chunk,
-                [(x._name, i) if isinstance(x, _Frame) else x for x in args],
-                chunk_kwargs,
-            )
-            for i in range(npartitions)
-        }
+    # Start with combined layers and dependencies for dfs
+    layers = dfs[0].dask.layers.copy()
+    dependencies = dfs[0].dask.dependencies.copy()
+    key_dependencies = (dfs[0].dask.key_dependencies or {}).copy()
+    for df in dfs[1:]:
+        layers.update(df.dask.layers)
+        dependencies.update(df.dask.dependencies)
+        key_dependencies.update(df.dask.key_dependencies or {})
+
+    # Blockwise Chunk Layer
+    chunk_name = f"{token or funcname(chunk)}-chunk-{token_key}"
+    layers[chunk_name] = partitionwise_graph(
+        chunk,
+        chunk_name,
+        *args,
+        **chunk_kwargs,
+    )
+    dependencies[chunk_name] = {df._name for df in dfs}
```

Comment:
Yeah - Using `map_partitions` was certainly my first thought, but I had a bit of trouble making it work (and using `partitionwise_graph` directly worked out of the box). After playing a bit more with something close to your suggestions, it seems to "mostly" work. However, using `list` arguments for the aggregation will lead to the injection of an extra `MaterializedLayer` (from what is interpreted as a `Delayed` dependency).  Still trying to figure out exactly what is goin on here and if there is a simple way to avoid the problem...

Reply:
> However, using `list` arguments for the aggregation

I'm not following, which arguments are lists?

That's too bad it didn't work out of the box, but I think there must be a way to get it working, and it would be worth the huge reduction in code.

Reply:
It looks like we just need an option to avoid normalizing the arguments (i.e. converting them to delayed) in `map_partitions`.

Reply:
What's getting converted to delayed? Is it happening in https://github.com/dask/dask/blob/27a7928fb750d604dfbad0b3f4361aa64a2c97b4/dask/array/core.py#L434-L453

Reply:
Marking this as "resolved", since the discussion effectively [moved here](https://github.com/dask/dask/pull/8468#discussion_r771719399)